agent/ebs/watcher_test.go

```diff
@@ -0,0 +1,281 @@
+// permissions and limitations under the License.
+
+package ebs
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/session/testconst"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/attachmentinfo"
+	apiebs "github.com/aws/amazon-ecs-agent/ecs-agent/api/resource"
+	mock_ebs_discovery "github.com/aws/amazon-ecs-agent/ecs-agent/api/resource/mocks"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/status"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	resourceAttachmentARN = "arn:aws:ecs:us-west-2:123456789012:attachment/a1b2c3d4-5678-90ab-cdef-11111EXAMPLE"
+	containerInstanceARN  = "arn:aws:ecs:us-west-2:123456789012:container-instance/a1b2c3d4-5678-90ab-cdef-11111EXAMPLE"
+	taskARN               = "task1"
+	taskClusterARN        = "arn:aws:ecs:us-west-2:123456789012:cluster/customer-task-cluster"
+	deviceName            = "/dev/xvdba"
+	volumeID              = "vol-1234"
+)
+
+type TestGroup struct {
+	wg     sync.WaitGroup
+	cancel context.CancelFunc
+}
+
+// NewTestGroup creates a TestGroup with the given cancellation function.
+func NewTestGroup(cancel context.CancelFunc) TestGroup {
+	return TestGroup{cancel: cancel}
+}
+
+// Start the watcher in a new goroutine. The watcher must be initialized with
+// the group's context.
+func (tg *TestGroup) Start(watcher *EBSWatcher) {
+	tg.wg.Add(1)
+	go func() {
+		defer tg.wg.Done()
+		watcher.Start()
+	}()
+}
+
+// Cancel the watcher started by Start() by calling the associated context and wait its completion.
+func (tg *TestGroup) Cancel() {
+	// Stop the watcher.
+	tg.cancel()
+
+	// Make sure all watcher have been stopped.
+	tg.wg.Wait()
+}
+
+func setupWatcher(ctx context.Context, cancel context.CancelFunc, agentState dockerstate.TaskEngineState,
+	ebsChangeEvent chan<- statechange.Event, discoveryClient apiebs.EBSDiscovery) *EBSWatcher {
+	return &EBSWatcher{
+		ctx:             ctx,
+		cancel:          cancel,
+		agentState:      agentState,
+		ebsChangeEvent:  ebsChangeEvent,
+		discoveryClient: discoveryClient,
+		mailbox:         make(chan func(), 1),
+	}
+}
+
+func TestHandleEBSAttachment(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	testGroup := NewTestGroup(cancel)
+	defer testGroup.Cancel()
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	taskEngineState := dockerstate.NewTaskEngineState()
+	eventChannel := make(chan statechange.Event)
+
+	mockDiscoveryClient := mock_ebs_discovery.NewMockEBSDiscovery(mockCtrl)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	mockDiscoveryClient.EXPECT().ConfirmEBSVolumeIsAttached(deviceName, volumeID).Return(nil).MinTimes(1).
+		Do(func(deviceName, volumeID string) {
+			wg.Done()
+		}).
+		Return(nil).
+		MinTimes(1)
```

Comment:
Does this mean we expect two calls both returning `nil` but the first one calls the `wg.Done()`?

Reply:
Ah, good catch. Fixed.